### ISSUE TITLE
No result in Metabase for simple MongoDB queries

### ISSUE THREAD

**Issue by anki-code** [metabase/metabase#10741]
I'm trying to get count of rows in MongoDB (3.4.20) collection on fresh install of Metabase 0.33 and I've got no result (endless inscription "Doing science"). But if I do this query in mongo client I've get the result immediately.

CASE 1 - Try to get simple count of rows

Steps:
1. First of all I try to do it in mongo client:
```
> db.Activities.count()
### IMMEDIATELY ###
8046328
```

2. Try to do it in Metabase and got no result:
![image](https://user-images.githubusercontent.com/1708680/63852440-f2fb0c00-c9a0-11e9-92b5-88554ed7d4a1.png)

3. Let's try to `explain` the query to calculate count (the query was built by MB):
```
> db.Activities.explain().aggregate([{"$group":{"_id":null,"count":{"$sum":1}}},{"$sort":{"_id":1}},{"$project":{"_id":false,"count":true}}])
{
        "stages" : [
                {
                        "$cursor" : {
                                "query" : {

                                },
                                "queryPlanner" : {
                                        "plannerVersion" : 1,
                                        "namespace" : "analytics.Activities",
                                        "indexFilterSet" : false,
                                        "parsedQuery" : {

                                        },
                                        "winningPlan" : {
                                                "stage" : "COLLSCAN",
                                                "direction" : "forward"
                                        },
                                        "rejectedPlans" : [ ]
                                }
                        }
                },
                {
                        "$group" : {
                                "_id" : {
                                        "$const" : null
                                },
                                "count" : {
                                        "$sum" : {
                                                "$const" : 1
                                        }
                                }
                        }
                },
                {
                        "$sort" : {
                                "sortKey" : {
                                        "_id" : 1
                                }
                        }
                },
                {
                        "$project" : {
                                "_id" : false,
                                "count" : true
                        }
                }
        ],
        "ok" : 1
}
```
No index. No result.

4. Let's add useless `match` clause (the data was started since 2018 but we `match` it since 2015):
```
> db.Activities.explain().aggregate([{"$match":{"EntityCreatedAt":{"$gt":ISODate("2015-08-27T17:37:07.592+03:00")}}},{"$group":{"_id":null,"count":{"$sum":1}}},{"$project":{"_id":false,"count":true}}])
{
        "stages" : [
                {
                        "$cursor" : {
                                "query" : {
                                        "EntityCreatedAt" : {
                                                "$gt" : ISODate("2015-08-27T14:37:07.592Z")
                                        }
                                },
                                "queryPlanner" : {
                                        "plannerVersion" : 1,
                                        "namespace" : "analytics.Activities",
                                        "indexFilterSet" : false,
                                        "parsedQuery" : {
                                                "EntityCreatedAt" : {
                                                        "$gt" : ISODate("2015-08-27T14:37:07.592Z")
                                                }
                                        },
                                        "winningPlan" : {
                                                "stage" : "COUNT_SCAN",
                                                "keyPattern" : {
                                                        "EntityCreatedAt" : 1
                                                },
                                                "indexName" : "EntityCreatedAt_1",
                                                "isMultiKey" : false,
                                                "multiKeyPaths" : {
                                                        "EntityCreatedAt" : [ ]
                                                },
                                                "isUnique" : false,
                                                "isSparse" : false,
                                                "isPartial" : false,
                                                "indexVersion" : 2,
                                                "indexBounds" : {
                                                        "startKey" : {
                                                                "EntityCreatedAt" : ISODate("2015-08-27T14:37:07.592Z")
                                                        },
                                                        "startKeyInclusive" : false,
                                                        "endKey" : {
                                                                "EntityCreatedAt" : ISODate("0NaN-NaN-NaNTNaN:NaN:NaNZ")
                                                        },
                                                        "endKeyInclusive" : true
                                                }
                                        },
                                        "rejectedPlans" : [ ]
                                }
                        }
                },
                {
                        "$group" : {
                                "_id" : {
                                        "$const" : null
                                },
                                "count" : {
                                        "$sum" : {
                                                "$const" : 1
                                        }
                                }
                        }
                },
                {
                        "$project" : {
                                "_id" : false,
                                "count" : true
                        }
                }
        ],
        "ok" : 1
}
```
Indexes. Works for 10 seconds.

5. If I try to get just N rows from the collection it works as expected.

**Comment by anki-code** [comment]
```
Open metabase main page

08-28 21:14:09 DEBUG middleware.log :: POST /api/dataset 200 [ASYNC: canceled] 21.0 s (6 DB calls) Jetty threads: 3/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:13 DEBUG middleware.log :: GET /api/user/current 200 1.6 ms (3 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:13 DEBUG middleware.log :: GET /api/session/properties 200 2.8 ms (1 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:13 DEBUG middleware.log :: GET /api/database 200 8.8 ms (3 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:13 DEBUG middleware.log :: GET /api/collection/root/items 200 3.6 ms (7 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:13 DEBUG middleware.log :: GET /api/setting 200 672.7 µs (0 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:13 DEBUG middleware.log :: GET /api/collection/root 200 1.1 ms (2 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0

Click Ask question, Simple, collection Activities

08-28 21:14:22 DEBUG middleware.log :: GET /api/database 200 10.2 ms (5 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:27 DEBUG middleware.log :: GET /api/table/32/fks 200 4.8 ms (5 DB calls) Jetty threads: 5/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0
08-28 21:14:27 DEBUG middleware.log :: GET /api/table/32/query_metadata 200 20.6 ms (8 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (118 total active threads) Queries in flight: 0

Click run and get first rows

08-28 21:14:30 DEBUG middleware.log :: POST /api/dataset 200 [ASYNC: completed] 626.5 ms (15 DB calls) Jetty threads: 3/50 (3 idle, 0 queued) (117 total active threads) Queries in flight: 0

Click Summarize - Count of rows - Done - Doing science

NOTHING

Wait 2 minutes

NOTHING

```
Browser Firefox 68.0.2

**Comment by anki-code** [comment]
CASE 2 - Two queries. Try to get simple group statistic for last 2 days by hands and in Metabase.

TEST 1 - MongoDB Console - It works for ONE MINUTE and USE INDEXES
```
> db.Activities.explain().aggregate([{$match:{"EntityCreatedAt":ISODate("2019-08-27T00:00:00.000+00:00"),}},{$group:{_id:{"EntityCreatedAt":{$dateToString:{format:"%Y-%m-%d",date:"$EntityCreatedAt"}}},count:{$sum:1}}},{$sort:{_id:1}}])
{
        "stages" : [
                {
                        "$cursor" : {
                                "query" : {
                                        "EntityCreatedAt" : ISODate("2019-08-27T00:00:00Z")
                                },
                                "fields" : {
                                        "EntityCreatedAt" : 1,
                                        "_id" : 0
                                },
                                "queryPlanner" : {
                                        "plannerVersion" : 1,
                                        "namespace" : "analytics.Activities",
                                        "indexFilterSet" : false,
                                        "parsedQuery" : {
                                                "EntityCreatedAt" : {
                                                        "$eq" : ISODate("2019-08-27T00:00:00Z")
                                                }
                                        },
                                        "winningPlan" : {
                                                "stage" : "PROJECTION",
                                                "transformBy" : {
                                                        "EntityCreatedAt" : 1,
                                                        "_id" : 0
                                                },
                                                "inputStage" : {
                                                        "stage" : "IXSCAN",
                                                        "keyPattern" : {
                                                                "EntityCreatedAt" : 1
                                                        },
                                                        "indexName" : "EntityCreatedAt_1",
                                                        "isMultiKey" : false,
                                                        "multiKeyPaths" : {
                                                                "EntityCreatedAt" : [ ]
                                                        },
                                                        "isUnique" : false,
                                                        "isSparse" : false,
                                                        "isPartial" : false,
                                                        "indexVersion" : 2,
                                                        "direction" : "forward",
                                                        "indexBounds" : {
                                                                "EntityCreatedAt" : [
                                                                        "[new Date(1566864000000), new Date(1566864000000)]"
                                                                ]
                                                        }
                                                }
                                        },
                                        "rejectedPlans" : [
                                                {
                                                        "stage" : "PROJECTION",
                                                        "transformBy" : {
                                                                "EntityCreatedAt" : 1,
                                                                "_id" : 0
                                                        },
                                                        "inputStage" : {
                                                                "stage" : "IXSCAN",
                                                                "keyPattern" : {
                                                                        "EntityCreatedAt" : 1,
                                                                        "ActivityCreatedAt" : 1
                                                                },
                                                                "indexName" : "EntityCreatedAt_1_ActivityCreatedAt_1",
                                                                "isMultiKey" : false,
                                                                "multiKeyPaths" : {
                                                                        "EntityCreatedAt" : [ ],
                                                                        "ActivityCreatedAt" : [ ]
                                                                },
                                                                "isUnique" : false,
                                                                "isSparse" : false,
                                                                "isPartial" : false,
                                                                "indexVersion" : 2,
                                                                "direction" : "forward",
                                                                "indexBounds" : {
                                                                        "EntityCreatedAt" : [
                                                                                "[new Date(1566864000000), new Date(1566864000000)]"
                                                                        ],
                                                                        "ActivityCreatedAt" : [
                                                                                "[MinKey, MaxKey]"
                                                                        ]
                                                                }
                                                        }
                                                },
                                                {
                                                        "stage" : "PROJECTION",
                                                        "transformBy" : {
                                                                "EntityCreatedAt" : 1,
                                                                "_id" : 0
                                                        },
                                                        "inputStage" : {
                                                                "stage" : "IXSCAN",
                                                                "keyPattern" : {
                                                                        "EntityCreatedAt" : 1,
                                                                        "ActivityCreatedAt" : 1,
                                                                        "TeamId" : 1,
                                                                        "SourceType" : 1
                                                                },
                                                                "indexName" : "EntityCreatedAt_1_ActivityCreatedAt_1_TeamId_1_SourceType_1",
                                                                "isMultiKey" : false,
                                                                "multiKeyPaths" : {
                                                                        "EntityCreatedAt" : [ ],
                                                                        "ActivityCreatedAt" : [ ],
                                                                        "TeamId" : [ ],
                                                                        "SourceType" : [ ]
                                                                },
                                                                "isUnique" : false,
                                                                "isSparse" : false,
                                                                "isPartial" : false,
                                                                "indexVersion" : 2,
                                                                "direction" : "forward",
                                                                "indexBounds" : {
                                                                        "EntityCreatedAt" : [
                                                                                "[new Date(1566864000000), new Date(1566864000000)]"
                                                                        ],
                                                                        "ActivityCreatedAt" : [
                                                                                "[MinKey, MaxKey]"
                                                                        ],
                                                                        "TeamId" : [
                                                                                "[MinKey, MaxKey]"
                                                                        ],
                                                                        "SourceType" : [
                                                                                "[MinKey, MaxKey]"
                                                                        ]
                                                                }
                                                        }
                                                }
                                        ]
                                }
                        }
                },
                {
                        "$group" : {
                                "_id" : {
                                        "EntityCreatedAt" : {
                                                "$dateToString" : {
                                                        "format" : "%Y-%m-%d",
                                                        "date" : "$EntityCreatedAt"
                                                }
                                        }
                                },
                                "count" : {
                                        "$sum" : {
                                                "$const" : 1
                                        }
                                }
                        }
                },
                {
                        "$sort" : {
                                "sortKey" : {
                                        "_id" : 1
                                }
                        }
                }
        ],
        "ok" : 1
}
```

TEST 2 - Metabase - NO INDEXES - NO RESULT
1. Go to metabase and set:
* filter - last 2 days
* groupby - CreatedAt by day

```
> db.Activities.explain().aggregate([{"$project":{"ActivityCreatedAt~~~day":{"$let":{"vars":{"column":"$ActivityCreatedAt"},"in":{"___date":{"$dateToString":{"format":"%Y-%m-%d","date":"$$column"}}}}}}},{"$match":{"ActivityCreatedAt~~~day":{"$gte":{"___date":"2019-08-26"},"$lte":{"___date":"2019-08-27"}}}},{"$project":{"_id":"$_id","___group":{"ActivityCreatedAt~~~day":"$ActivityCreatedAt~~~day"}}},{"$group":{"_id":"$___group","count":{"$sum":1}}},{"$sort":{"_id":1}},{"$project":{"_id":false,"ActivityCreatedAt~~~day":"$_id.ActivityCreatedAt~~~day","count":true}},{"$sort":{"ActivityCreatedAt~~~day":1}}])
{
        "stages" : [
                {
                        "$cursor" : {
                                "query" : {

                                },
                                "fields" : {
                                        "ActivityCreatedAt" : 1,
                                        "_id" : 1
                                },
                                "queryPlanner" : {
                                        "plannerVersion" : 1,
                                        "namespace" : "analytics.Activities",
                                        "indexFilterSet" : false,
                                        "parsedQuery" : {

                                        },
                                        "winningPlan" : {
                                                "stage" : "COLLSCAN",
                                                "direction" : "forward"
                                        },
                                        "rejectedPlans" : [ ]
                                }
                        }
                },
                {
                        "$project" : {
                                "_id" : true,
                                "ActivityCreatedAt~~~day" : {
                                        "$let" : {
                                                "vars" : {
                                                        "column" : "$ActivityCreatedAt"
                                                },
                                                "in" : {
                                                        "___date" : {
                                                                "$dateToString" : {
                                                                        "format" : "%Y-%m-%d",
                                                                        "date" : "$$column"
                                                                }
                                                        }
                                                }
                                        }
                                }
                        }
                },
                {
                        "$match" : {
                                "ActivityCreatedAt~~~day" : {
                                        "$gte" : {
                                                "___date" : "2019-08-26"
                                        },
                                        "$lte" : {
                                                "___date" : "2019-08-27"
                                        }
                                }
                        }
                },
                {
                        "$project" : {
                                "_id" : "$_id",
                                "___group" : {
                                        "ActivityCreatedAt~~~day" : "$ActivityCreatedAt~~~day"
                                }
                        }
                },
                {
                        "$group" : {
                                "_id" : "$___group",
                                "count" : {
                                        "$sum" : {
                                                "$const" : 1
                                        }
                                }
                        }
                },
                {
                        "$sort" : {
                                "sortKey" : {
                                        "_id" : 1
                                }
                        }
                },
                {
                        "$project" : {
                                "_id" : false,
                                "count" : true,
                                "ActivityCreatedAt~~~day" : "$_id.ActivityCreatedAt~~~day"
                        }
                },
                {
                        "$sort" : {
                                "sortKey" : {
                                        "ActivityCreatedAt~~~day" : 1
                                }
                        }
                }
        ],
        "ok" : 1
}
```

**Comment by mazameli** [comment]
Hey, @anki-code! Long time no see. I've been having trouble reproducing this. Is this happening with all your mongo collections, or just a particular one?

**Comment by anki-code** [comment]
Hi @mazameli! )
Yep I long time with you and already use MB with PostgreSQL/Greenplum and now I've tried to use MB for simple case on MongoDB. This case I've got for collections with many rows, 8 millions of documents for example.  

The collection has indexes on `EntityCreatedAt` and I expect fast response (like in mongo client) but I caught sight that MB made queries without indexes and as result we have no result at minutes or hours also. MongoDB [don't use indexes for group](https://stackoverflow.com/questions/20455752/mongodb-aggregation-framework-does-group-use-index) by (only for match).

I've added the `explain` result for MB query to CASE 1 and 2. Take a look and you'll get the problem - full scan of documents.
